### PR TITLE
CA-359869: Make Sysfs.list robust against disappearing devices

### PR DESCRIPTION
**Pull request by robhoes**

### CODE REVIEW

**Review by lindig — APPROVED** [review]

File: ocaml/networkd/lib/network_utils.ml [hunk 1]
```diff
@@ -198,7 +198,9 @@ module Sysfs = struct
         false
 
   let list () =
-    let is_dir name = Sys.is_directory ("/sys/class/net/" ^ name) in
+    let is_dir name =
+      try Sys.is_directory ("/sys/class/net/" ^ name) with _ -> false
+    in
     Sys.readdir "/sys/class/net"
     |> Array.to_list
     |> List.filter (fun name -> is_dir name && is_ether_device name)
```
Comment (lindig) [inline ocaml/networkd/lib/network_utils.ml:206]: An alternative would be to check using `Sys.file_exists` in the predicate in addition. `file_exists` is true for directories, too.

Comment (robhoes) [reply ocaml/networkd/lib/network_utils.ml:206]: I think here we want _only_ directories.

**Review by lindig — APPROVED** [review]

File: ocaml/networkd/lib/network_utils.ml (hunk 1 above)
Comment (lindig) [reply ocaml/networkd/lib/network_utils.ml:206]: My suggestion was to check for: exists && is directory && is_ether_devive.

Comment (robhoes) [reply ocaml/networkd/lib/network_utils.ml:206]: Oh you said _in addition_ (to the directory check)...

**Review by edwintorok — APPROVED** [review]
I see that almost all callers of read_one_line need to wrap it in a try/with. Would it be better if read_one_line returned a Result instead to force the caller into handling the error condition?
In fact isn't `is_dir` redundant here? If it is not a dir then trying to open a file inside it will raise ENOENT and cause is_ether_device to return false anyway.

**Comment by lindig** [comment]
>  In fact isn't `is_dir` redundant here? If it is not a dir then trying to open a file inside it will raise ENOENT and cause is_ether_device to return false anyway.

Good observation. Historically `is_ether_device` was introduced after `is_dir`.

**Comment by robhoes** [comment]
Yes, strictly speaking it appears to be redundant (now). However, it may be good to have the explicit check there just in case someone "optimises" or "refactors" it later for some reason.

**Pull request merged**